sklearn/metrics/cluster/unsupervised.py

```diff
@@ -210,6 +210,16 @@ def silhouette_samples(X, labels, metric='euclidean', **kwds):
 
     """
     X, labels = check_X_y(X, labels, accept_sparse=['csc', 'csr'])
+
+    # Check for diagonal entries in precomputed distance matrix
+    if metric == 'precomputed':
+        diag_indices = np.diag_indices(X.shape[0])
+        if np.any(X[diag_indices]):
```

Comment:
Maybe we still need to consider floating point error here? e.g., ``np.any([1e-18]) is True``. We can use ``np.isclose``.

Reply:
What's the purpose? Is it to allow some small but non-zero values on the diagonal?

That seems counter to the pull request because tiny values will change the result of the silhouette score.

Reply:
The point here is to avoid unexpected error, we generally avoid doing things like `float == 0``.

> tiny values will change the result of the silhouette score.

Do you have an example to show that small numbers on the diagonal (e.g., 1e-10) will change the score significantly?

Reply:
> The point here is to avoid unexpected error

Uhhhhh, errors are generally unexpected, it's pretty rare that you aim to write code with errors.

> Do you have an example to show that small numbers on the diagonal (e.g., 1e-10) will change the score significantly?

How much is "significant"? To me any answer that is not exactly the same (up to floating point precision limits) is not good enough.

Reply:
> Uhhhhh, errors are generally unexpected, it's pretty rare that you aim to write code with errors.

I mean sometimes some elements on the diagonal are actually zero, but due to floating point errors, we need to use np.isclose to check whether they're zero, see e.g., #10005 #11591

Reply:
Perhaps there is a misunderstanding about this change? The code does not check if values are equal to zero. It checks if they are non-zero. So anything that is not an exact 0 will cause an error. This forces users to zero the diagonal before calling ``silhouette_samples``.

Reply:
Perhaps there is a misunderstanding about this change? The code does not check if values are equal to zero. It checks if they are non-zero. Yes, I know that these are logically equivalent modulo ``not`` but I think there is a difference in doctrine. I would prefer that we force users to zero the diagonal before calling ``silhouette_samples`` instead of tolerating some "likely to be intended to be zero but due to FP precision issues are not exactly zero" values as these can then potentially cause minor changes to the silhouette score.

Reply:
@sjtrny ,

1. you can just use `np.diagonal` to directly get the values
2. As noted by @qinhanmin2014 	, exact comparison with float values are almost always false, due to numerical precision issues. Having a small tolerance threshold would make sense here.

Also, ideally, I wouldn't call reviews from long-time core-devs a "waste of time".

Reply:
2. I want it to return False if it is not exactly zero since these numerical precision issues cause incorrect calculation. I want to force users (myself) to set the diag to zero before calling this function to avoid incorrect calculation.

Keeping that in mind it does not make sense to tolerate some values that are close to but not zero.

I did not call the review a waste of time. I was referring to the effort in implementing it and deciding on threshold value since it is at odds of the goal of this PR. I highly value the work and effort of all contributors.

Reply:
We understand what you want to do.

The point is that very small non-zero values may be present in the pre-computed matrix, and maybe these small non-zeros values have **no significant effect on the silhouette score**. If that's the case, then the check is a bit too strict.

We basically need you to illustrate your claim "tiny values will change the result of the silhouette score". You can do this e.g. by computing the score when setting the diagonal to zero vs setting the diagonal to random values with e.g. `|x| < 1e-10`.

In any case, I would suggest:

- address previous comments
- indicate in the docstring for `metric` that the matrix must have zero diagonal entries when it's precomputed
- indicate in the error message how to set the diagonal entries to zero (e.g. `np.fill_diagonal(X, 0)`)